src/MongoDB.Driver/MongoCredential.cs

```diff
@@ -536,6 +560,14 @@ private static MongoCredential FromComponents(string mechanism, string source, s
                 default:
                     throw new NotSupportedException(string.Format("Unsupported MongoAuthenticationMechanism {0}.", mechanism));
             }
+
+            void ensureNullOrExternal(string value)
```

Comment:
I think this might not be a good scenario for a local function for a couple reasons:

1. One parameter (mechanism) is implied/captured and the other (value) is passed in
2. The value parameter should really be called source but can't because that would be a name conflict

I think in  this case a normal function will work better:

    private static void EnsureNullOrExternalSource(string mechanism, string source)
    {
        if (source != null && source != "$external")
        {
            throw new ArgumentException($"A {mechanism} source must be $external.", nameof(source));
        }
    }

Reply:
Done